File: examples/canonical-analytics/rust/src/analytics/distributions.rs
use super::{match_date, match_multi, LatencyByService, LatencyHistogram};
use crate::data::models::LatencySample;
use crate::filters::FilterState;
use std::collections::HashMap;

pub fn percentile(values: &mut [f64], p: f64) -> f64 {
    if values.is_empty() {
        return 0.0;
    }
    values.sort_by(|a, b| a.partial_cmp(b).unwrap());
    let idx = ((values.len() as f64 * p / 100.0) as usize).min(values.len() - 1);
    values[idx]
}

pub fn aggregate_latency_histogram(
    data: &[LatencySample],
    filters: &FilterState,
) -> LatencyHistogram {
    if data.is_empty() {
        return LatencyHistogram::default();
    }

    let mut values = Vec::new();
    for sample in data {
        if !match_date(&sample.timestamp, &filters.date_from, &filters.date_to) {
            continue;
        }
        if !match_multi(&sample.service, &filters.services) {
            continue;
        }
        if !match_multi(&sample.environment, &filters.environment) {
            continue;
        }
        values.push(sample.latency_ms);
    }

    if values.is_empty() {
        return LatencyHistogram::default();
    }

    let mut sorted_vals = values.clone();
    let p50 = percentile(&mut sorted_vals, 50.0);
    let p95 = percentile(&mut sorted_vals, 95.0);
    let p99 = percentile(&mut sorted_vals, 99.0);

    LatencyHistogram {
        values,
        p50,
        p95,
        p99,
    }
}

pub fn aggregate_latency_by_service(
    data: &[LatencySample],
    filters: &FilterState,
) -> LatencyByService {
    if data.is_empty() {
        return LatencyByService::default();
    }

    let mut by_srv: HashMap<String, Vec<f64>> = HashMap::new();
    for sample in data {
        if !match_date(&sample.timestamp, &filters.date_from, &filters.date_to) {
            continue;
        }
        if !match_multi(&sample.service, &filters.services) {
            continue;
        }
        if !match_multi(&sample.environment, &filters.environment) {
            continue;
        }
        by_srv
            .entry(sample.service.clone())
            .or_default()
            .push(sample.latency_ms);
    }

    let mut services = Vec::new();
    let mut p50s = Vec::new();
    let mut p95s = Vec::new();
    let mut p99s = Vec::new();

    let mut keys: Vec<String> = by_srv.keys().cloned().collect();
    keys.sort();

    for k in keys {
        let mut vals = by_srv.get(&k).unwrap().clone();
        services.push(k);
        p50s.push(percentile(&mut vals, 50.0));
        p95s.push(percentile(&mut vals, 95.0));
        p99s.push(percentile(&mut vals, 99.0));
    }

    LatencyByService {
        services,
        p50s,
        p95s,
        p99s,
    }
}

File: examples/canonical-analytics/elixir/analytics_workbench/lib/analytics_workbench/analytics/distributions.ex
defmodule AnalyticsWorkbench.Analytics.Distributions do
  alias AnalyticsWorkbench.Filters

  def aggregate_latency_histogram(samples, %Filters{} = filters) do
    filtered = apply_filters(samples, filters)
    values = Enum.map(filtered, & &1.latency_ms) |> Enum.sort()

    %{
      values: values,
      p50: percentile(values, 50),
      p95: percentile(values, 95),
      p99: percentile(values, 99)
    }
  end

  def aggregate_latency_by_service(samples, services, %Filters{} = filters) do
    filtered = apply_filters(samples, filters)

    service_stats = Enum.map(services, fn s ->
      s_samples = Enum.filter(filtered, & to_string(&1.service) == s.name)
      values = Enum.map(s_samples, & &1.latency_ms) |> Enum.sort()
      {s.name, percentile(values, 50), percentile(values, 95), percentile(values, 99)}
    end)

    %{
      services: Enum.map(service_stats, & elem(&1, 0)),
      p50s: Enum.map(service_stats, & elem(&1, 1)),
      p95s: Enum.map(service_stats, & elem(&1, 2)),
      p99s: Enum.map(service_stats, & elem(&1, 3))
    }
  end

  defp percentile([], _), do: 0.0
  defp percentile(values, p) do
    idx = floor(length(values) * p / 100)
    idx = min(idx, length(values) - 1)
    Enum.at(values, idx) |> to_float() |> Float.round(2)
  end

  defp to_float(val) when is_integer(val), do: val / 1
  defp to_float(val), do: val

  defp apply_filters(samples, filters) do
    samples
    |> filter_by_multi(:service, filters.services)
    |> filter_by_multi(:environment, filters.environment)
  end

  defp filter_by_multi(events, _key, []), do: events
  defp filter_by_multi(events, key, values) do
    Enum.filter(events, &(to_string(Map.get(&1, key)) in values))
  end
end
